sourcegraph/sourcegraph
# Wildcard [Line Chart]: Add chart keyboard navigation

Closes https://github.com/sourcegraph/sourcegraph/issues/44046

## Background
This PR implements most of the logic from this proposal about chart (widget-like) keyboard navigation ([see this document](https://docs.google.com/document/d/1rF4jietYrAVxm9gOLFhP2xlQYQZp7rX3oVdD5SI0iuo/edit#heading=h.2ix5bz3tt0z5)). TL:DR this PR implements arrow keys navigation over line chart datapoints. See a quick demo below. 


https://user-images.githubusercontent.com/18492575/204672491-586fd5e8-bc98-43b3-ac4e-e7e6af70bd46.mp4


## Open questions
- What is the best short cut trigger to enter the keyboard navigation mode? Is it any arrow key press or it should be more designated like press Enter to enter the keyboard navigation mode? (at the moment we ask to press Enter to enter the arrow keys navigation mode)
- Do we need to move focus to the card itself when we exit chart keyboard navigation mode? or we could move focus to the svg root element? (at the moment we return focus to the svg element)


## Test plan

- Check that you can navigate through all points on the chart (all points are reachable from the keyboard)
- Check that keyboard navigation works properly on the code insight charts (the biggest consumer of the chart at the moment in sourcegraph) 



## App preview:

- [Web](https://sg-web-vk-add-chart-keyboard-navigation.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.
